security/tokensource/token_source.go

```diff
@@ -0,0 +1,147 @@
+package tokensource
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path"
+	"sync"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/netcracker/qubership-core-lib-go/v3/logging"
+)
+
+const (
+	serviceAccountDir = "/var/run/secrets/kubernetes.io/serviceaccount"
+	secretsDir        = "/var/run/secrets"
+)
+
+var (
+	logger   = logging.GetLogger("token-file-storage")
+	mu       sync.RWMutex
+	launched = make(map[string]*fileTokenSource)
+)
+
+func GetToken(ctx context.Context, audience string) (string, error) {
+	mu.RLock()
+	tokenSource, ok := launched[audience]
+	mu.RUnlock()
+	if ok {
+		return tokenSource.Token()
+	}
+
+	mu.Lock()
+	defer mu.Unlock()
+	tokenSource, ok = launched[audience]
+	if ok {
+		return tokenSource.Token()
+	}
+
+	entries, err := os.ReadDir(secretsDir)
+	if err != nil {
+		return "", fmt.Errorf("failed to get entries of dir %s: %w", secretsDir, err)
+	}
+	for _, entry := range entries {
+		if entry.Name() != audience {
+			continue
+		}
+		ts, err := newFileTokenSource(ctx, fmt.Sprintf("%s/%s/token", secretsDir, audience))
+		if err != nil {
+			return "", fmt.Errorf("failed to create a tokensource for token with audience %s: %w", audience, err)
+		}
+		launched[audience] = ts
+		return ts.Token()
+	}
+	return "", fmt.Errorf("token with audience %s not found in %s", audience, secretsDir)
+}
+
+type fileTokenSource struct {
+	mu       sync.RWMutex
+	err      error
+	token    string
+	tokenDir string
+}
+
+func NewDefault(ctx context.Context) (*fileTokenSource, error) {
+	return New(ctx, serviceAccountDir)
+}
+
+func New(ctx context.Context, tokenDir string) (*fileTokenSource, error) {
+	if tokenDir == "" {
+		return nil, fmt.Errorf("tokenDir is an empty string, use NewDefault if default service account dir needed or specify tokenDir")
+	}
+
+	ts := &fileTokenSource{
+		tokenDir: tokenDir,
+	}
+	err := ts.refreshToken()
+	if err != nil {
+		return nil, err
+	}
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize file watcher: %w", err)
+	}
+	err = watcher.Add(ts.tokenDir + "/")
+	if err != nil {
+		return nil, fmt.Errorf("failed to add path %s to file watcher: %w", ts.tokenDir, err)
```

Comment:
There is resource leakage. Look at the NewWatcher internals. You need to call `Close()` on watcher instance